README.md
# Mautic K8s

**Deploy Mautic in a scalable Kubernetes cluster.**

This project is an _opinionated_ way to deploy Mautic using:

- **Helm 3 charts** to manage deploying Mautic 2.x
- **nginx-proxy** as our ingress
- **PHP-FPM** with PHP 7.3
- **Nginx** on each application server to work with php-fpm
- **PVC on EFS** for Mautic `media`, `spool`, `cache`, and `logs`.
- **EFK** in cluster
- **Redis** for shared php sessions (but we hope to use this for another Cache plugin soon)
- **RabbitMQ**
- **GitLab CI/CD** with multidev branches
- **RDS** for managed MariaDB

### Notes on Opinionated Mautic Setup

- We make `local.php` read-only with an empty array and require developers to configure everything in `parameters_local.php` where we insert environment based switches.
- We use specialized bash scripts to programmatically link `vendor` and `bin` directories, along with `/mautic_overrides` which are specified in a relative folder path to the `/mautic` project. 

## Requirements

The following are current requirements to work with this project, but considering the interoperability of Helm Charts, we plan to support other Kubernetes platforms and configurations in the future. 

- [AWS Account](https://aws.amazon.com/) - To deploy to AWS EKS.
- [GitLab](https://gitlab.com/) - To leverage CI/CD to deploy. 
- [Helm CLI](https://helm.sh/) - To manage your Kubernetes instance.
- [Lando](https://lando.dev/) - To develop locally.

## Supported K8s Backplanes

- [x] AWS EKS - Amazon Elastic Container Service for Kubernetes
- [ ] GKE - Google Kubernetes Engine
- [ ] AKS - Azure Kubernetes Service
- [ ] Digital Ocean Kubernetes
- [ ] Rancher Kubernetes

_[Open an issue](https://github.com/FacetInteractive/mautic-k8s/issues/new) if you plan to work on one of the unchecked Kubernetes providers, we'd love to know!_

# Current Major Initiatives

- [x] Upgrade to Mautic 2.16.4
- [x] Upgrade to PHP 7.3
- [x] Confirm Redis Connection in PHP 7.3 for Shared Sessions
- [ ] Copy/backfill dev databases as new dev* environments are provisioned
- [ ] Branch to support Mautic 3.x / PHP 7.4

# Get Started Mautic-K8s

1. Fork this repo
2. Copy the following files for your project's configurations:

```bash
cp infra/k8s/values.default.yaml infra/k8s/values.yaml

cp mautic_override/app/config/local.default.php mautic_override/app/config/local.php

cp mautic_override/app/config/parameters_local.default.php mautic_override/app/config/parameters_local.php
```

# Local Mautic Dev With Lando

For instructions on how to set up `mautic-k8s` with Lando, [see the documentation here](./wiki/local-dev-getting-started.md)

# K8s Setup

For instructions on how to deploy `mautic-k8s` for the first time, [see the documentation here](./wiki/k8s.md)

## Environment Variables

The following environment variables must be set in GitLab CI or an equivalent CI pipeline are below.

**NOTE**: To get a better understanding of how these variables are used you will have to follow the daisy chain of these environment variables from `.gitlab-ci.yml` to `infra/k8s/values.default.yaml` to `infra/k8s/templates/mautic_statefulset.yaml`


|Key                            |Description                    |Example                            |Environment|
|-------------------------------|-------------------------------|-----------------------------------|-----------|
|DEV_APP_SECRET_KEY             |Mautic Application Secret Key  |                                   |DEV|
|DEV_GA_TRACKING_ID             |Google Analytics Tracking Code |UA-123456789-1                     |DEV|
|DEV_INGRESS_DOMAIN             |Ingress Domain for K8s         |k8s.mydomain.com                   |DEV|
|DEV_MAILER_HOST                |Amazon SES Host (full URL)     |email-smtp.us-east-1.amazonaws.com |DEV|
|DEV_MAILER_REGION              |Amazon Mailer Region (full URL)|email-smtp.us-east-1.amazonaws.com |DEV|
|DEV_MAILER_SECRET_KEY          |Amazon SES IAM Secret Key      |awsIamSecretKey                    |DEV|
|DEV_MAILER_USER                |Amazon SES IAM User            |awsIamUser                         |DEV|
|DEV_SAML_IDP_IDENTITY_ID       |Typically the hostname of your marketing domain|https://www.mydomain.com|DEV|
|ECR_REGISTRY                   |Subdomain for your ECR Registry|1234567890.dkr.ecr.us-east-1.amazonaws.com|DEV/PROD|
|MAUTIC_CORS_DOMAINS            |Cross-Origin Resource Sharing, name the domains you want to allow to send tracking pixel and form requests to your Mautic instance. Accepts comma-separated values.|https://mydomain.com\,https://www.mydomain.com\,https://store.mydomain.com|DEV/PROD|
|MAUTIC_REPOSITORY_URL          |URL for your PHP-FPM Image in your registry.|1234567890.dkr.ecr.us-east-1.amazonaws.com/mautic-php-fpm|DEV/PROD|
|NGINX_REPOSITORY_URL           |URL for your Nginx Image in your registry.|1234567890.dkr.ecr.us-east-1.amazonaws.com/mautic-nginx|DEV/PROD|
|PROD_APP_SECRET_KEY            |Mautic Application Secret Key  |                                   |DEV|
|PROD_GA_TRACKING_ID            |Google Analytics Tracking Code |UA-123456789-1                     |DEV|
|PROD_INGRESS_DOMAIN            |Ingress Domain for K8s. Typically the TLD if mautic production is a subdomain|mydomain.com|DEV|
|PROD_MAILER_REGION             |Amazon Mailer Region (full URL)|email-smtp.us-east-1.amazonaws.com |DEV|
|PROD_MAILER_SECRET_KEY         |Amazon SES IAM Secret Key      |awsIamSecretKey                    |DEV|
|PROD_MAILER_USER               |Amazon SES IAM User            |awsIamUser                         |DEV|
|PROD_SAML_IDP_IDENTITY_ID      |Typically the hostname of your marketing domain|https://www.mydomain.com|DEV|
|RDS_HOST                       |Amazon RDS Hostname (URL)      |instance.hashString.us-east-1.rds.amazonaws.com|DEV/PROD|
|RDS_PASSWORD                   |Amazon RDS Password            |password                           |DEV/PROD|
|RDS_ROOT_PASSWORD              |Amazon RDS Root Password       |rootPassword                       |DEV/PROD|
|REGISTRY_AWS_REGION            |Amazon Region for ECR          |us-east-1                          |DEV/PROD|
|TRUSTED_PROXIES                |Upstream trusted proxies IP range|10.0.0.0/8\,172.16.0.0/12        |DEV/PROD|

_**Comma Separated Values must be escaped**_

## Environment Variables - _@TODO_ 

1. In the future it would be better to use GitLab environment stages instead of switching our variable naming convention.
2. This variable set and instructions are opinionated towards AWS SES. It would be great if we can set up a conditional statement for the `.Values.project.mailerHost` based on the selection of a transactional email service. 
3. Refactor `ECR_REGISTRY` to optionally work with any registry. 

# Contributors

[Facet Interactive](https://facetinteractive.com/services/mautic-development-managed-services?utm_source=mautic-k8s&utm_medium=github&utm_campaign=README.md) leads and sponsors the `mautic-k8s` project to support enterprise deployments of Mautic on Kubernetes. 

& special thanks to: 

- [Axelerant](https://axelerant.com/?utm_source=mautic-k8s&utm_medium=github&utm_campaign=README.md) for Helmizing Mautic and setting up GitLab CI/CD.

## Contributing

If you would like to get involved, please: 

* **Open an Issue** to let others know what you plan to work on.
* **Join the [#kubernetes](https://mautic.slack.com/archives/C01G6LHLM5M) channel in Mautic Slack**. 
* **Submit a Pull Request**. Don't forget to update the CHANGELOG.md with a summary of your changes. 

# License

MIT License

See [LICENSE](./LICENSE) to see the full text.

Important Note:

Mautic K8s as a distribution software is licensed under MIT License but bundled components can use different licenses and may overwrite the terms and conditions detailed in MIT License. Mautic itself is licensed under GPLv3 [mautic/mautic/LICENSE.txt](https://github.com/mautic/mautic/blob/features/LICENSE.txt) and Mautic is a registered trademark of David Hurley.  



CHANGELOG.md
# Changelog

## 2021-xx - Upcoming 

- ? 

## 2021-02-17 - Cron Fixes

- RabbitMQ requires separate cron for processing email_hit and page_hit.

## 2021-02-04 - OSS Prep & Project Organization

- Move all relevant infrastructure-as-code (IaC) configuration into the `/infra` folder.
- Move `post_deploy.sh` into the scripts folder and make some `@TODO`s for future improvements to make it agnostic of k8s. 
- Update the `README` to better introduce the open source project.
- Move all variables to GitLab CI. No variables should live within the project. 
- Update `/wiki` to use a better naming convention for various pages.
- Update `/wiki` to relevant current project information.
- Add environment variables table to README. 

## 2021-01-14 - Critical Mautic Security 2021-01-14

- Patch applied. 

## 2021-01-13 - Lando LAMP Configuration

- While working with recent updates to Lando containers, we started seeing an error related to `security.limit_extensions` and couldn't resolve it.
- Brought parity to nginx-lando.conf and symfony.conf, but still needs some work. 
- Implemented Lando LAMP recipe as a configuration to keep development moving. These aren't interchangeable with the K8s deployment, however we're not experiencing these issues on our production Docker images.
- Fix for `scripts/pull_live_db.sh` to allow for gzip compression.
- Fix for `scripts/pull_live_db.sh` to stop passing `--databases` flag which creates a `USE databasename;` statement and messes up imports elsewhere.
- Fix for `scripts/reset_admin_password.sh` to escape characters into mysql query.

## 2020-08-24 - Lando LEMP Configuration

Lando requires a custom `nginx.conf` in order to work in a LEMP stack. 

While implementing this a number of other Mautic configuration improvements were made. 

- Compared a copy of Drupal8 Lando Recipe `nginx.conf` and this gist for Mautic: https://gist.github.com/proffalken/ebfa9debc4eef929b0163d11a80af349
- `Nginx-lando.conf` has parity with our `symfony.conf` settings. In the future it would be better if both of these configs were maintained in the same file.
- Added `php.ini` which was sourced from `mautic-eb` open source project. This `php.ini` is not yet implemented in our production deployments. 
- Implemented a `config_override.php` to stream logs to `/dev/stdout` so we can pick them up via Docker Logs / Lando Logs.
- Added `rabbitmq` configuration to Lando services.
- Added `redis` configuration to Lando services.
- Set up `TRUSTED_PROXIES` as an environment variable, and Lando does not require a proxy.
- Remove `/mautic/vendor` and `/mautic/bin` from the git history
- Clean up Dockerfile-base to remove `--enable-gd-native-ttf` flag which is no longer necessary in PHP7.2.
- Remove `xdebug` from `Dockerfile-base`.
- Preparing for future programmatic composer builds of Mautic, copy the `parameters_local.php` to a `config/` folder. Eventually we will symlink this file or copy it at build time. 
- `reset_admin_password.sh` to Reset Admin Password for Local on lando build.
    - Lando shortcut: `lando reset-admin-pw`
- `scripts/sync_db_live_to_local.sh` to remotely download a copy of the MySQL Database.
    - Lando shortcut: `lando pull-live-db` 

### WIP Commits

- WIP: Added conditional check for Lando host domain DEV_HOST so that local development is done with dev mode turned on automatically. 

### Deployment Notes

- Requires `TRUSTED_PROXIES` to be defined in the environment conditionally on builds.

### New @ToDos

- [ ] Use one nginx.conf for both Lando local development and K8s deployments.
- [ ] Use `php.ini` in K8s Deployments
- [ ] Rebuild the Dockerfile-base with `from php7.3-base` instead of custom source. 
- [x] Lando Logs vs. Docker Logs - Determine if we should conditionally check for a LANDO environment before streaming to `/dev/stdout`
- [ ] Allow `reset-admin-pw` to take an argument.


api-script/bv-100.py
#! /usr/bin/env python3
import requests

base_url = "https://mautic.k8s.facetinteractive.com/api"
username = "jordan"
passwd = "XXXXX"


def get_contact_data(contact_id: int):
    url = f"{base_url}/contacts/" + str(contact_id)
    r = requests.get(url, auth=(username, passwd))
    return r.json()

def get_contact_by_email(email):
    url = f"{base_url}/contacts"
    r = requests.get(url, auth=(username, passwd), params={"search":email})
    if r.status_code == 200:
        emails = r.json()["emails"]
        return int(list(emails)[0])

def create_contact(contact):
    url = f"{base_url}/contacts/new"
    r = requests.post(url, auth=(username, passwd), data=contact)
    return r.json()

def get_email_id(slug):
    url = f"{base_url}/emails"
    r = requests.get(url, auth=(username, passwd), params={"search":slug})
    if r.status_code == 200:
        emails = r.json()["emails"]
        return int(list(emails)[0])

def send_email_to_contact(contact_id, email_id, tokens):
    url = f"{base_url}/emails/{email_id}/contact/{contact_id}/send"
    r = requests.post(url, auth=(username, passwd), data=tokens)
    return r

def get_segments():
    url = f"{base_url}/segments"
    r = requests.get(url, auth=(username, passwd), params={"limit": 100})
    return r.json()

"""
>>> list(r["lists"])
['20', '21', '31', '47', '12', '64', '51', '43', '59', '14', '52', '53', '46', '37', '36', '16', '13', '15', '18', '66', '11', '28', '34', '40', '32', '33', '8', '29', '27', '30', '54', '55', '56', '57', '65', '63', '62', '61', '10', '26', '17', '22', '24', '25', '19', '60', '38', '39', '23', '67', '35', '41', '42', '58', '48', '49', '50', '44', '45']
>>> r["lists"]['67']
{'isPublished': True, 'dateAdded': '2020-01-20T15:57:52+00:00', 'dateModified': None, 'createdBy': 1, 'createdByUser': 'Jordan Ryan', 'modifiedBy': None, 'modifiedByUser': None, 'id': 67, 'name': 'Test Segment', 'alias': 'test-segment', 'description': None, 'filters': [], 'isGlobal': True, 'isPreferenceCenter': False}
"""


def add_contact_to_segment(segment_id, contact_id):
    url = f"{base_url}/segments/{segment_id}/contact/{contact_id}/add"
    r = requests.post(url, auth=(username, passwd))
    return r
    

get_contact_data(79814)


contact = {
    "firstname": "Lakshmi",
    "lastname": "P",
    "email":"lakshmi.narasimhan@axelerant.com",
    "company":"Axl",
    "phone": "+91 875 447 8359",
    "city": "Chennai",
    "state": "Tamil Nadu",
    "zipcode": "600061",
    "country": "India",
    "twitter": "lakshminp",
}
    

create_contact(contact)

get_email_id("bv-100")

data = {"tokens" : {"lead.customer.first_name" : "FOOBAR"} }
send_email_to_contact(80592, 194, data)

add_contact_to_segment(67, 80592)
